src/Perl6/Actions.nqp

```diff
@@ -9716,6 +9710,72 @@ class Perl6::Actions is HLL::Actions does STDActions {
                 }
             }
 
+            if %info<signature_constraint> {
+                my $var-qast := QAST::Var.new( :name($name), :scope('local') );
+                my $var-decont := get_decont_name()
+                                    ?? QAST::Var.new( :name(get_decont_name()), :scope('local') )
+                                    !! QAST::Op.new( :op('decont'), $var-qast );
+                my $sigc_name := QAST::Node.unique('__lowered_sig_constraint_');
+                my $sigc-var := QAST::Var.new( :name($sigc_name), :scope<local>);
+                my $sigc-qast;
+                # Produce different code for generic/non-generic signatures because in the latter case instantiation
+                # code would be a waste of memory and performance.
+                if %info<signature_constraint>.is_generic {
+                    $sigc-qast := QAST::Op.new(
+                                    :op<if>,
+                                    QAST::Op.new(
+                                        :op<callmethod>,
+                                        :name<is_generic>,
+                                        QAST::Op.new(
+                                            :op<bind>,
+                                            QAST::Var.new( :name($sigc_name), :scope<local>, :decl<var>),
+                                            QAST::Op.new(
+                                                :op<getattr>,
+                                                $inst_param,
+                                                QAST::Op.new(:op<what>, $inst_param),
+                                                QAST::SVal.new(:value('$!signature-constraint'))))),
+                                    QAST::Op.new(
+                                        :op<callmethod>,
+                                        :name<instantiate_generic>,
+                                        $sigc-var,
+                                        QAST::Op.new(
+                                            :op<ctxlexpad>,
+                                            QAST::Op.new(:op<ctx>))),
+                                    $sigc-var );
+                }
+                else {
+                    $sigc-qast := QAST::Op.new(
+                                    :op<getattr>,
+                                    $inst_param,
+                                    QAST::Op.new(:op<what>, $inst_param),
+                                    QAST::SVal.new(:value('$!signature-constraint')));
+                }
+                $var.push(QAST::ParamTypeCheck.new(
+                    QAST::Op.new(
+                        # If argument is a type object and is the same as parameter default then skip signature
+                        # matching. So far, this is the best way I know to determine if corresponding argument was
+                        # passed or not without inspecting the capture which is too slow.
```

Comment:
Inspecting captures and matching types is what new-disp does all the time and efficiently too thanks to caching and guards. Sounds like this would be an opportunity for utilizing the new possibilities.

So instead of generating code into each routine, the dispatcher should contain the logic for handling arguments with signatures.

Reply:
@niner  Unfortunately, I'm not into dispatchers yet and will need time to learn them sufficiently to start actually doing something. The whole signature binding could be done as a dispatcher, I think. But before this is done I'd rather have a temporary fix in place. This would also allow to merge roast tests, so when dispatched solution arrives we have them in place to ensure it works as intended.

Aside of this, I think the whole thing is going to be different with RakuAST anyway. It is likely that dispatching will come as part of RakuAST implementation.

So, if there are no other objections, I'd be for merging this solution for now. As I'm involuntarily getting more and more involved into signature-related stuff, perhaps, when time allows, I'd be the one to start rewriting it. :)

Reply:
Fair enough :) Please, go ahead